### PR TITLE
Updated starknet_api_openrpc to support code autogeneration

### PR DESCRIPTION
**Pull request by phklive**
Autogeneration of typings would be of real utility to the entirety of the ecosystem from the open-rpc specs of Starknet. Unfortunately the specs had wrongly been created ( Ex. title was used as description, results were optionals, etc...). 
Hence I have updated as much as i could to correctly add titles, correct return types, correct return options, and others.

With this PR we can now use open-rpc-typings to create auto generated types for rust.

Previous behavior: 
Code auto-generation was broken and outputed types with random titles.
Also, trace and write are have errors in them hence code auto-generation is not possible.
We will need to re-write them and fix bugs to enable it.

New behavior: 
The auto generated code is correct and usable

IMO: 
1st step - Make open-rpc functional with code auto-generation
2nd step - Improve this functional open-rpc
3rd step - Re-write and update to newer version ( api / trace / write )

How to use this pr: 
`git clone repo`
`npm install @open-rpc/typings`
`open-rpc-typing -d ./api/starknet_api_openrpc.json --output-rs .`

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/84)

### CODE REVIEW

**Review by ArielElp — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/84)* status: 0 of 2 files reviewed, 9 unresolved discussions (waiting on @phklive)

___
*[`api/starknet_api_openrpc.json` line 19 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSehxUXCte__XaOkF2F:-NSehxUXCte__XaOkF2G:b-envqcs) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L19)):*
> ```JSON
>                     "required": true,
>                     "schema": {
>                         "title": "Block id",
> ```

Are titles in schemas defined by a reference required? I think that if BLOCK_ID has a title then this reference doesn't require title duplication (same elsewhere)
___
*[`api/starknet_api_openrpc.json` line 19 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSeilCWDwrjGqvNv-Mq:-NSeilCWDwrjGqvNv-Mr:bn6oo1p) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L19)):*
> ```JSON
>                     "required": true,
>                     "schema": {
>                         "title": "Block id",
> ```

Do titles with spaces work with auto generations? e.g. in return results for some methods you used underscores rather than spaces, we should be consistent here
___
*[`api/starknet_api_openrpc.json` line 469 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSejjlm4fZn1B4EUay9:-NSejjln-f8RtQadYZpT:br686v5) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L469)):*
> ```JSON
>                 "schema": {
>                     "type": "array",
>                     "title": "Field element",
> ```

CallResult? 



_Code quote:_
```JSON
Field element
```
___
*[`api/starknet_api_openrpc.json` line 778 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSekQZU7HSoaMDEvK7f:-NSekQZU7HSoaMDEvK7g:b-yle2qv) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L778)):*
> ```JSON
>             },
>             "EMITTED_EVENT": {
>                 "title": "Emitted event",
> ```

Choosing this example but true for all other instances, do we go with EmittedEvent? I would guess spaces mess up generation



_Code quote:_
```JSON
Emitted event
```
___
*[`api/starknet_api_openrpc.json` line 787 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSekJnV3XFRt6_2pbIq:-NSekJnV3XFRt6_2pbIr:byyyqrg) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L787)):*
> ```JSON
>                     },
>                     {
>                         "title": "Event emission",
> ```

Not a very clear name, how about EventContext?



_Code quote:_
```JSON
Event emission
```
___
*[`api/starknet_api_openrpc.json` line 820 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSezLr9AbAZJg_cFL8V:-NSezLr9AbAZJg_cFL8W:b-997em9) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L820)):*
> ```JSON
>                 "allOf": [
>                     {
>                         "title": "Params",
> ```

I would expect generation tools to handle allOf types, have a separate title for every parts looks weird IMO. In any case, Params isn't saying much, I suggest "EventEmitter" or something of the sort.



_Code quote:_
```JSON
Params
```
___
*[`api/starknet_api_openrpc.json` line 887 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSf--EU4E-vrR_dGhSM:-NSf--EV5ycOmk-2CBzE:bubn1xn) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L887)):*
> ```JSON
>                         "type": "array",
>                         "items": {
>                             "title": "Keys",
> ```

does an array item need a title or can this be removed?
___
*[`api/starknet_api_openrpc.json` line 1339 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSf-LNr3F2GEXabUbam:-NSf-LNr3F2GEXabUban:bfj8m72) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L1339)):*
> ```JSON
>                     {
>                         "type": "object",
>                         "title": "Params",
> ```

same allOf comment (is it needed for generation), plus I would avoid Params as a title
___
*[`api/starknet_api_openrpc.json` line 2731 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSf0-mVBXaGmPn99ov7:-NSf0-mVBXaGmPn99ov8:bax6y81) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L2731)):*
> ```JSON
>                 ]
>             },
>             "FEE_ESTIMATE": {
> ```

Missing required?

**Review by phklive — COMMENTED** [review]
I will fix what need to be fixed, indeed names can be improved to be more explicit, some titles are required.

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-:-NSg6jxm30Pp4--wE9fv:b-hq2hx8)* status: 0 of 2 files reviewed, 9 unresolved discussions (waiting on @ArielElp)

___
*[`api/starknet_api_openrpc.json` line 19 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSeilCWDwrjGqvNv-Mq:-NSg3BRaCN-TsmGYMAwB:baeuo46) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L19)):*
<details><summary><i>Previously, ArielElp wrote…</i></summary><blockquote>

Do titles with spaces work with auto generations? e.g. in return results for some methods you used underscores rather than spaces, we should be consistent here
</blockquote></details>

Titles with spaces do work, I will update the starknet_ with spaces to be consistent.
___
*[`api/starknet_api_openrpc.json` line 19 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSehxUXCte__XaOkF2F:-NSg3LMM9dOZfoC-pVlz:b-uga1ym) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L19)):*
<details><summary><i>Previously, ArielElp wrote…</i></summary><blockquote>

Are titles in schemas defined by a reference required? I think that if BLOCK_ID has a title then this reference doesn't require title duplication (same elsewhere)
</blockquote></details>

Unfortunately the name that is used is not the field that is considered by the open-rpc playground or the code auto-generation, hence without the "title" field the title field is missing in the open-rpc playground and titles are auto generated in the types hence creating unusable types.

As follows: 

___
*[`api/starknet_api_openrpc.json` line 469 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSejjlm4fZn1B4EUay9:-NSg49F-1c9MrxljCuxk:bys4ju4) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L469)):*
<details><summary><i>Previously, ArielElp wrote…</i></summary><blockquote>

CallResult? 
</blockquote></details>

In the Starknet specs right now, starknet_call is returning an array of Field elements, as follows:


I have just added the title field to annotate and help with code-autogeneration and playground.
___
*[`api/starknet_api_openrpc.json` line 778 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSekQZU7HSoaMDEvK7f:-NSg4VTD13vPyo0xLhJd:b-dglixg) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L778)):*
<details><summary><i>Previously, ArielElp wrote…</i></summary><blockquote>

Choosing this example but true for all other instances, do we go with EmittedEvent? I would guess spaces mess up generation
</blockquote></details>

Spaces or underscores do not mess up code-autogeneration hence i have chosen to use spaces as a rule in all titles, imo titles should be human readable.
___
*[`api/starknet_api_openrpc.json` line 787 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSekJnV3XFRt6_2pbIq:-NSg4gpMAQZHbCfFrmRW:b-lfg17f) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L787)):*
<details><summary><i>Previously, ArielElp wrote…</i></summary><blockquote>

Not a very clear name, how about EventContext?
</blockquote></details>

Will change to event context 👍
___
*[`api/starknet_api_openrpc.json` line 820 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSezLr9AbAZJg_cFL8V:-NSg4rO7CKjhQ2LdXtMb:b-amj080) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L820)):*
<details><summary><i>Previously, ArielElp wrote…</i></summary><blockquote>

I would expect generation tools to handle allOf types, have a separate title for every parts looks weird IMO. In any case, Params isn't saying much, I suggest "EventEmitter" or something of the sort.
</blockquote></details>

Code-autogeneration doesn't handle allOf types, it generates types as follows starting with "AllOf":

___
*[`api/starknet_api_openrpc.json` line 887 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSf--EU4E-vrR_dGhSM:-NSg5H-oCG130Pq1ExQX:b-czc5fu) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L887)):*
<details><summary><i>Previously, ArielElp wrote…</i></summary><blockquote>

does an array item need a title or can this be removed?
</blockquote></details>

I will make sure they are absolutely necessary if not they will be removed, but through my experience they did help with code-autogeneration.
___
*[`api/starknet_api_openrpc.json` line 1339 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSf-LNr3F2GEXabUbam:-NSg5RHd3AeVUQFlrEtJ:b-662y09) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L1339)):*
<details><summary><i>Previously, ArielElp wrote…</i></summary><blockquote>

same allOf comment (is it needed for generation), plus I would avoid Params as a title
</blockquote></details>

title is required, what title would you recommend?
___
*[`api/starknet_api_openrpc.json` line 2731 at r1](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-NSf0-mVBXaGmPn99ov7:-NSg6gM08HSWfUdoAFVX:bo6apac) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/a2439715bba2d175c5252b0d0a197019ecdca41a/api/starknet_api_openrpc.json#L2731)):*
<details><summary><i>Previously, ArielElp wrote…</i></summary><blockquote>

Missing required?
</blockquote></details>

Indeed.

**Comment by phklive** [comment]
Also, trace and write are have errors in them hence code auto-generation is not possible.
We will need to re-write them and fix bugs to enable it.

**Comment by phklive** [comment]
Also here are examples of the files auto-generated before and after pr:

before: https://gist.github.com/phklive/58a2cacb33eae9d0b66fbfcb1fd03b93

after: https://gist.github.com/phklive/1b81127fce118f30872850dda899ef8e

File: package.json [hunk 1]
```diff
@@ -16,4 +16,4 @@
         "fs-extra": "10.1.0",
         "@json-schema-tools/dereferencer": "1.5.4"
     }
-}
+}
```
Comment (drspacemn) [inline package.json:19]: nit: newline EOF

File: api/starknet_api_openrpc.json [hunk 2]
```diff
@@ -2199,120 +2555,176 @@
...
                         "type": "string"
                     },
                     "size": {
+                        "title": "Size",
                         "type": "integer",
                         "minimum": 1
                     },
                     "members": {
                         "type": "array",
+                        "title": "Struct member",
```
[the remark this answers is not in the document]
Comment (phklive) [reply api/starknet_api_openrpc.json:2577]: After fix this is the new generated type: 
SCR-20230414-lkjv

File: api/starknet_api_openrpc.json [hunk 3]
```diff
@@ -2093,38 +2420,45 @@
...
                                 "type": "array",
+                                "title": "L1 handler",
                                 "items": {
                                     "$ref": "#/components/schemas/DEPRECATED_CAIRO_ENTRY_POINT"
                                 }
                             }
                         }
                     },
                     "abi": {
+                        "title": "Contaract ABI",
```
[the remark this answers is not in the document]
Comment (phklive) [reply api/starknet_api_openrpc.json:2461]: Fixed
SCR-20230414-lkyq

**Review by ArielElp — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-:-NTHlMXYBqygQPPSkuFK:bnfp4nl)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @drspacemn, @greged93, and @phklive)

**Comment by phklive** [comment]
> :lgtm:
> 
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-:-NTHlMXYBqygQPPSkuFK:bnfp4nl)_ status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @drspacemn, @greged93, and @phklive)

Great 🚀

**Comment by drspacemn** [comment]
Is there a way to avoid this generation?
https://gist.github.com/phklive/1b81127fce118f30872850dda899ef8e#file-index-rs-L935

**Comment by phklive** [comment]
> Is there a way to avoid this generation? https://gist.github.com/phklive/1b81127fce118f30872850dda899ef8e#file-index-rs-L935

Hey, unfortunately this is the only name i wasn't able to fix, it's globally created from all the method names and there doesn't seem to be a place to add a global title that would enable this auto-generated placeholder to be correctly named.

I should have pointed it out earlier before you found it.

**Review by ArielElp — APPROVED** [review]
> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/84)* status: all files reviewed, 3 unresolved discussions (waiting on @greged93 and @phklive)

**Review by phklive — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-:-NTNnFZ-8BiAT64HeDB9:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @greged93)

**Comment by greged93** [comment]
There seems to still be one issue on the auto generation for `DeprecatedContractClass` and `ContractClass`:
both contain a field `entry_points_by_type ` which during auto generation maps to the same type (`EntryPointsByType`).  `DeprecatedContractClass` should `DeprecatedCairoEntryPoint` and not `SierraEntryPoint`. 
Links provided for clarity: 
- [Specs `CONTRACT_CLASS`](https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json#L2061)
- [Specs `DEPRECATED_CONTRACT_CLASS`](https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json#L2104)
- [Generated types](https://gist.github.com/phklive/1b81127fce118f30872850dda899ef8e#file-index-rs-L152)

**Comment by Mirko-von-Leipzig** [comment]
Yeah honestly the codegen tool (for rust) really is not great..

One of our team members attempted his own version [here](https://github.com/sergey-melnychuk/iamgroot) which seems a bit more sane.

**Comment by phklive** [comment]
> There seems to be an issue on the auto generation for `DeprecatedContractClass` and `ContractClass`: both contain a field `entry_points_by_type ` which during auto generation maps to the same type (`EntryPointsByType`). `DeprecatedContractClass` should have `DeprecatedCairoEntryPoint` for the type of `entry_points_by_type` and not `SierraEntryPoint`. Links provided for clarity:
> 
> * [Specs `CONTRACT_CLASS`](https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json#L2061)
> * [Specs `DEPRECATED_CONTRACT_CLASS`](https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json#L2104)
> * [Generated types](https://gist.github.com/phklive/1b81127fce118f30872850dda899ef8e#file-index-rs-L152)

Indeed here the code auto-generation is confused because both use similar names, hence during generation it does not know how to use the type.

The iamgroot code-autogeneration tool makes the same error except reversed:
SCR-20230419-lpse

The potential fix that I found would be to rename the titles to make the distinguishable from "entry points by type" to "deprecated entry points by type" and "constructor" to "deprecated constructor", etc... Which will not change the underlying types in the Starknet specs but only disambiguate those parameters enabling code auto-generation and imo unchanged or clearer understanding by the users.
SCR-20230419-lrln

**Comment by Mirko-von-Leipzig** [comment]
@sergey-melnychuk btw ^^.

One solution: namespace these inner types within modules, which at least prevents this legal-spec-but-doesn't-work-with-bad-tools dance.

It does make me skeptical that this official tool works at all though. Feels like one would have to audit every codegen every time.

**Comment by sergey-melnychuk** [comment]
@Mirko-von-Leipzig 

The [iamgroot](https://github.com/sergey-melnychuk/iamgroot) does not need such kind of fix because it simply works. The [generated code](https://github.com/sergey-melnychuk/iamgroot/blob/1259ba2111f83d63858b7b44f6b05c02d70841c4/examples/gen.rs#L1049) actually works against real endpoints, including proxy mode.

This PR just tries to optimize the spec for specific tool, which in my opinion makes no sense, as it is clearly an indicator of a problem with a tool, not a spec.

**Comment by phklive** [comment]
I am confused by your response, I am obviously not trying to optimize the specs for any type of auto-generation tool, I am just making sure that it generates correct code every time ( playground, code-generation, etc... ) by adding correct labels and correcting errors, the only thing I optimize is our Starknet-specs.

I understand that you have built iamgroot and I am sure that it's a great tool and that you did a great work but here as you can see in my previous response your tool had outputed incorrect code, and unfortunately after my fix too, which the official tool didn't.

I think iamgroot is good but we wouldn't be able to say that "it simply works", no software "simply works" every time.

**Comment by sergey-melnychuk** [comment]
@phklive 

Sorry, I got too defensive for no reason. I misread Mirko's comment as a suggestion of a fix for iamgroot.

I just ~wasted~ spent too much time on code-generation topic, trying out tools and even building one. What I've learnt is that if a spec isn't built with a code-gen in mind (this is mostly about naming types and avoiding anonymous ones), no tool can "fix" it into meaningful code.

**Comment by phklive** [comment]
Hey guys, I don't know what workflow is awaiting approval, what do we miss for a merge?

Thanks

**Review by ArielElp — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/84#-:-NUW3dQ8BO2zl8wmlj4p:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @greged93 and @phklive)

**Review by ArielElp — APPROVED** [review]
> Dismissed @greged93 from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/84)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @phklive)

**Pull request merged**